syne_tune/backend/python_backend/python_backend.py

```diff
@@ -0,0 +1,101 @@
+        should not reference any global variable or module and should have as arguments all the keys of `config_space`.
+        When deserializing, a md5 is checked to ensure consistency.
+
+        For instance, the following function is a valid way of defining a backend on top of simple function:
+
+        ```python
+        def f(x):
+            import logging
+            import time
+            from syne_tune import Reporter
+            root = logging.getLogger()
+            root.setLevel(logging.DEBUG)
+            reporter = Reporter()
+            for i in range(5):
+                reporter(step=i + 1, y=x + i)
+
+
+        from syne_tune.backend.python_backend.python_backend import PythonBackend
+        from syne_tune.config_space import uniform
+
+        config_space = {"x": uniform(-10, 10)}
+        backend = PythonBackend(tune_function=f, config_space=config_space)
+        ```
+        See `examples/launch_height_python_backend.py` for a full example.
+        :param tune_function: a python function to be tuned, the function should call Syne Tune reporter to report
+        metrics and be serializable, imports should be performed inside the function body.
+        :param config_space: config_space used to in Syne Tune, it must corresponds to key words arguments of
+        `tune_function`.
+        :param rotate_gpus: in case several GPUs are present, each trial is
+            scheduled on a different GPU. A new trial is preferentially
+            scheduled on a free GPU, and otherwise the GPU with least prior
+            assignments is chosen. If False, then all GPUs are used at the same
+            time for all trials.
+        :param delete_checkpoints: If True, checkpoints of stopped or completed
+            trials are deleted
+        """
+        super(PythonBackend, self).__init__(
+            entry_point=str(Path(__file__).parent / "python_entrypoint.py"),
+            rotate_gpus=rotate_gpus,
+            delete_checkpoints=delete_checkpoints,
+        )
+        self.config_space = config_space
+        # save function without reference to global variables or modules
+        self.tune_function = types.FunctionType(tune_function.__code__, {})
+        self.tune_function_path = self.local_path / "tune_function"
+
+    def set_path(self, results_root: Optional[str] = None, tuner_name: Optional[str] = None):
+        super(PythonBackend, self).set_path(results_root=results_root, tuner_name=tuner_name)
+        if self.local_path.exists():
+            print(f"path {self.local_path} already exists, make sure you have a unique tuner name.")
+            logging.error(f"path {self.local_path} already exists, make sure you have a unique tuner name.")
+        self.tune_function_path = self.local_path / "tune_function"
+
+    def _schedule(self, trial_id: int, config: Dict):
+        self.save_tune_function(self.tune_function)
```

Comment:
Why do we need to save the tune function every time we call schedule? Wouldn't it be enough to save it once in the init?

Reply:
I see your point, the reason of not doing it initially in the constructor is that the path can be changed after the creation of the backend, I added a check so that the operation is only performed once.